rfc/0001.md

```diff
@@ -0,0 +1,148 @@
+- Start Date: 2021-02-09
+- RFC PR: https://github.com/nens/threedi-modelchecker
+
+# Summary
+
+Add an API for reading from and writing to 3Di model schematisation (sqlite)
+files.
+
+# Basic examples
+
+Initialisation:
+
+```
+>>> from threedi_modelchecker import ThreediDatabase
+>>> schematisation = ThreediDatabase({"db_path": "/path/to/sqlite"})
+```
+
+Reading records: SQLAlchemy API
+```
+>>> channels = schematisation.channels
```

Comment:
Like the basic idea to wrap all `resources` together like this. Bonus would be to integrate validation in this as well, for example:

```
schematisation.validate()
schematisation.channels.validate()  # validate all channels
schematisation.channels.first().validate() # Not sure if this is possible/or we need to do this a bit different....
```

Reply:
I also like the idea of wrapping `resources` together! A while back I did an experiment with binding all checks (for validation) in a common resource, see https://github.com/nens/threedi-modelchecker/pull/22.

However, I wonder if wrapping all the SQLAlchemy models in a `Query` is useful. Basic queries might be handy but what if you want to make a bit more complex queries? For example how would you join two tables together? In this case you probably need to use the original defined SQLAlchemy models:
```python
from threedi_modelchecker import ThreediDatabase
from threedi_modelchecker.threedi_model import models

schematisation = ThreediDatabase({"db_path": "/path/to/sqlite"})
channels_with_conn_nodes = schematisation.channels.join(
    models.ConnectionNode, models.ConnectionNode.id == models.Channel.connection_node_start_id
)
```

Why not use the models directly:

```python
from threedi_modelchecker import ThreediDatabase
from threedi_modelchecker.threedi_model import models

schematisation = ThreediDatabase({"db_path": "/path/to/sqlite"})
session = schematisation.get_session()
channels_with_conn_nodes = session.query(models.Channel).join(
    models.ConnectionNode, models.ConnectionNode.id == models.Channel.connection_node_start_id
)
```
Isn't this more SQLAlchemy-like? 
I also don't understand where the `session` object lives in the first example. Shouldn't this be an external-object, as stated below?

Reply:
Good points. I am a novice SQLAlchemy users so I have no idea what is SQLAlchemy-like.

I thought that creating and destroying a session in one go would be possible with `schematisation.channels`; e.g. when the query goes out of scope, the session should go to. But no idea if that would work. Maybe best to rigidly stick to the SQLAlchemy patterns and not try to outsmart them. So I am :+1: on doing what you propose.

Reply:
I did some research. You really have to explicitly call `session.close()`. Else you will run into connection limits at some point. The SQLAlchemy docs are pretty clear on this subject:

> As a general rule, keep the lifecycle of the session separate and external from functions and objects that access and/or manipulate database data. This will greatly help with achieving a predictable and consistent transactional scope.

If some object needs to be closed, I think we should stick to the context manager pattern `with...as`. Precisely the second example. So I'll change also reading into that. Less nice API, but SQLAlchemy forces us into this.